Source/ProgressionSystemRuntime/Private/LevelActors/PSStarActor.cpp

```diff
@@ -162,29 +162,53 @@ void APSStarActor::OnInitialized(const FVector& PreviousActorLocation)
-		|| !ensureMsgf(StarMeshComponent, TEXT("ASSERT: [%i] %hs:\n'StarMeshComponent' is not valid!"), __LINE__, __FUNCTION__))
+	if (!ensureMsgf(StarMeshComponent, TEXT("ASSERT: [%i] %hs:\n'StarMeshComponent' is not valid!"), __LINE__, __FUNCTION__))
 	{
 		return; // Early return if pointers are invalid
 	}
 
+	UMaterialInstanceDynamic* StarProgressionDynamicMaterial = nullptr;
+
 	// locked stars
 	if (StarActorState == EPSStarActorState::Locked)
 	{
-		StarMeshComponent->SetMaterial(0, UPSDataAsset::Get().GetLockedProgressionMaterial());
+		const float LockedStar = 0.0f;
+		StarProgressionDynamicMaterial = UPSWorldSubsystem::Get().GetStarProgressionDynamicMaterial(EPSStarActorState::Locked);
+		SetStarActorProgressMeshMaterial(StarProgressionDynamicMaterial, LockedStar);
 		return;
 	}
 
 	// unlocked stars with fractional part
 	if (AmountOfStars > 0 && AmountOfStars < 1)
 	{
-		StarMeshComponent->SetMaterial(0, StarDynamicProgressMaterial);
-		StarDynamicProgressMaterial->SetScalarParameterValue(UPSDataAsset::Get().GetStarMaterialSlotName(), AmountOfStars / UPSDataAsset::Get().GetStarMaterialFractionalDivisor()); // StarMaterialFractionalDivisor is hardcoded value to 3 to tweak bad UV to simulate it's working
+		const float PartialUnLockedStarAmount = AmountOfStars / UPSDataAsset::Get().GetStarMaterialFractionalDivisor();
+		StarProgressionDynamicMaterial = UPSWorldSubsystem::Get().GetStarProgressionDynamicMaterial(EPSStarActorState::Partial); // StarMaterialFractionalDivisor is hardcoded value to 3 to tweak bad UV to simulate it's working
+		SetStarActorProgressMeshMaterial(StarProgressionDynamicMaterial, PartialUnLockedStarAmount);
 		return; // Early return for fractional stars
 	}
 
 	// unlocked stars EPSStarActorState::Unlocked
-	StarMeshComponent->SetMaterial(0, UPSDataAsset::Get().GetUnlockedProgressionMaterial());
+	const float UnLockedStarAmount = 1.0f;
+	StarProgressionDynamicMaterial = UPSWorldSubsystem::Get().GetStarProgressionDynamicMaterial(EPSStarActorState::Unlocked);
+	SetStarActorProgressMeshMaterial(StarProgressionDynamicMaterial, UnLockedStarAmount);
+}
+
+// Applies the star dynamic material
+void APSStarActor::SetStarActorProgressMeshMaterial(class UMaterialInstanceDynamic* StarDynamicMaterial, float StarProgressionAmount)
+{
+	if (!ensureMsgf(StarDynamicMaterial, TEXT("ASSERT: [%i] %hs:\n'StarDynamicMaterial' is not valid!"), __LINE__, __FUNCTION__))
+	{
+		return; // Early return if pointers are invalid
+	}
+
+	// Add a face texture over current star material
+	const FPSRowData& CurrentSettingsRowData = UPSWorldSubsystem::Get().GetCurrentProgressionSettingsRowByName();
+	const FName StarFaceTextureParameter = UPSDataAsset::Get().GetStarFaceTextureParameter();
+	UTexture* StarFaceTexture = CurrentSettingsRowData.StarFaceTexture;
+
+	StarMeshComponent->SetMaterial(0, StarDynamicMaterial);
```

Comment:
StarMeshComponent might crash as null, add ensure

Reply:
@moinm3uw It's not allowed to leave pointers unhandled in the current scope, even if they were validated earlier in the callstack.

In your case, since `StarMeshComponent` was already checked in a previous function (UpdateStarActorProgressMeshMaterial), you safely can validate pointer with `checkf` instead of `ensureMsgf`. This guarantees the pointer is never null at this point, as it has been validated earlier, otherwise controllable crash with checkf, to handle edge cases when this function was called in the wrong way, like by designers in blueprints

Reply:
got it,  done https://github.com/moinm3uw/ProgressionSystem/pull/55/commits/fae77d5835c95befc86093045528fc8a8e83859e